File: src/data/convert_excel_to_json.py
import pandas as pd
import json
import shutil
from datetime import datetime
from pathlib import Path

# Paths
data_dir = Path(r"c:\Users\maitr\.gemini\antigravity\scratch\antigravity-word-hoard\src\data")
excel_file = data_dir / "Liste refaite v7.xlsx"
json_file = data_dir / "words.json"
backup_file = data_dir / f"words_backup_{datetime.now().strftime('%Y-%m-%d_%H-%M-%S')}.json"

print(f"📂 Reading Excel file: {excel_file}")

# Read Excel file
df = pd.read_excel(excel_file)

print(f"✅ Loaded {len(df)} rows and {len(df.columns)} columns")
print(f"\n📋 Columns found:")
for i, col in enumerate(df.columns, 1):
    print(f"  {i}. {col}")

# Create backup of current words.json
if json_file.exists():
    print(f"\n💾 Creating backup: {backup_file.name}")
    shutil.copy(json_file, backup_file)

# Convert DataFrame to list of dictionaries
# Replace NaN with empty strings for consistency
df = df.fillna("")

# Convert to JSON-serializable format
words_data = df.to_dict(orient='records')

# Write to JSON file with proper formatting
print(f"\n💾 Writing to {json_file}")
with open(json_file, 'w', encoding='utf-8') as f:
    json.dump(words_data, f, ensure_ascii=False, indent=2)

print(f"\n✅ Conversion complete!")
print(f"   - Total words: {len(words_data)}")
print(f"   - Backup saved: {backup_file.name}")
print(f"   - New file: {json_file.name}")

# Show sample of first word
if words_data:
    print(f"\n📝 Sample (first word):")
    sample = words_data[0]
    for key, value in list(sample.items())[:10]:  # Show first 10 fields
        print(f"   {key}: {value}")
    if len(sample) > 10:
        print(f"   ... and {len(sample) - 10} more fields")

File: src/data/convert_excel_to_json.js
import XLSX from 'xlsx';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Paths
const dataDir = __dirname;
const excelFile = path.join(dataDir, 'Liste refaite v7.xlsx');
const jsonFile = path.join(dataDir, 'words.json');
const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
const backupFile = path.join(dataDir, `words_backup_${timestamp}.json`);

console.log(`📂 Reading Excel file: ${excelFile}`);

// Read Excel file
const workbook = XLSX.readFile(excelFile);
const sheetName = workbook.SheetNames[0];
const worksheet = workbook.Sheets[sheetName];

// Get raw data as array of arrays
const rawData = XLSX.utils.sheet_to_json(worksheet, { header: 1, defval: '', raw: false });

console.log(`✅ Total rows: ${rawData.length}`);

// First row should be headers
const headers = rawData[0];
console.log(`\n📋 Headers (row 1):`, headers);

// Rest are data rows
const dataRows = rawData.slice(1);

// Convert to array of objects
const data = dataRows.map(row => {
    const obj = {};
    headers.forEach((header, index) => {
        // Use header as key, or skip if empty
        if (header && header.trim() !== '') {
            obj[header] = row[index] || '';
        }
    });
    return obj;
});

console.log(`\n✅ Converted ${data.length} data rows`);

// Show columns
if (data.length > 0) {
    console.log(`\n📋 Columns in output:`);
    const columns = Object.keys(data[0]);
    columns.forEach((col, i) => {
        console.log(`  ${i + 1}. ${col}`);
    });
}

// Create backup of current words.json
if (fs.existsSync(jsonFile)) {
    console.log(`\n💾 Creating backup: ${path.basename(backupFile)}`);
    fs.copyFileSync(jsonFile, backupFile);
}

// Write to JSON file
console.log(`\n💾 Writing to ${path.basename(jsonFile)}`);
fs.writeFileSync(jsonFile, JSON.stringify(data, null, 2), 'utf-8');

console.log(`\n✅ Conversion complete!`);
console.log(`   - Total words: ${data.length}`);
console.log(`   - Backup saved: ${path.basename(backupFile)}`);
console.log(`   - New file: ${path.basename(jsonFile)}`);

// Show sample of first word
if (data.length > 0) {
    console.log(`\n📝 Sample (first word):`);
    const sample = data[0];
    const keys = Object.keys(sample);
    keys.forEach(key => {
        const value = sample[key];
        if (value !== '') {
            console.log(`   ${key}: ${value}`);
        }
    });
}
